### ISSUE TITLE
return %{...} fails to compile

### ISSUE THREAD

**Issue by Papierkorb** [opal/opal#1089]
Hi,
The following snippet breaks on line 6:

````ruby
def fine
  %{a}
end

def broken
  return %{a} ### Here
end
````

The compile error is as follows:
````
SyntaxError: An error occurred while compiling: (file)

unexpected '"tRCURLY"'
Source: (file):6:13
  return %{a}
~~~~~~~~~~~~^
````

Interestingly enough `return %q{a}` works fine, it's just the `%{a}` form that fails if used with a return statement.

Thank you for opal!

**Comment by alexdowad** [comment]
Using my same debugging process again... first edit build command for parser to add `-t` flag, rebuild parser, then:

```
22:07 ~prog/Ruby/opal % cat test.rb                                            [master]
$:.unshift Dir.pwd+'/lib'
require 'opal'
p = Opal::Parser.new
p.instance_eval { @yydebug = true }
p p.parse('return %{a}')
22:07 ~prog/Ruby/opal % ruby test.rb                                           [master]
read    :kRETURN(kRETURN) ["return", [1, 0]]

shift   kRETURN

        [ (kRETURN ["return", [1, 0]]) ]
goto    22
        [ 0 22 ]

read    :tPERCENT(tPERCENT) ["%", [1, 7]]

reduce  kRETURN --> primary
        [ (primary (:return)) ]

goto    42
        [ 0 42 ]

reduce  primary --> arg
        [ (arg (:return)) ]

goto    19
        [ 0 19 ]

shift   tPERCENT
        [ (arg (:return)) (tPERCENT ["%", [1, 7]]) ]

goto    226
        [ 0 19 226 ]

read    :tLBRACE(tLBRACE) ["{", [1, 8]]

shift   tLBRACE
        [ (arg (:return)) (tPERCENT ["%", [1, 7]]) (tLBRACE ["{", [1, 8]]) ]

goto    54
        [ 0 19 226 54 ]

read    :tIDENTIFIER(tIDENTIFIER) ["a", [1, 9]]

shift   tIDENTIFIER
        [ (arg (:return)) (tPERCENT ["%", [1, 7]]) (tLBRACE ["{", [1, 8]]) (tIDENTIFIER 
["a", [1, 9]]) ]

goto    84
        [ 0 19 226 54 84 ]

read    :tRCURLY(tRCURLY) ["}", [1, 10]]

reduce  tIDENTIFIER --> variable
        [ (arg (:return)) (tPERCENT ["%", [1, 7]]) (tLBRACE ["{", [1, 8]]) (variable (:identifier, :a)) ]

goto    209
        [ 0 19 226 54 209 ]

reduce  variable --> var_ref
        [ (arg (:return)) (tPERCENT ["%", [1, 7]]) (tLBRACE ["{", [1, 8]]) (var_ref (:call, nil, :a, (:arglist))) ]

goto    49
        [ 0 19 226 54 49 ]

reduce  var_ref --> primary
        [ (arg (:return)) (tPERCENT ["%", [1, 7]]) (tLBRACE ["{", [1, 8]]) (primary (:call, nil, :a, (:arglist))) ]

goto    42
        [ 0 19 226 54 42 ]

reduce  primary --> arg
        [ (arg (:return)) (tPERCENT ["%", [1, 7]]) (tLBRACE ["{", [1, 8]]) (arg (:call, nil, :a, (:arglist))) ]

goto    249
        [ 0 19 226 54 249 ]

reduce  arg --> arg_value
        [ (arg (:return)) (tPERCENT ["%", [1, 7]]) (tLBRACE ["{", [1, 8]]) (arg_value (:call, nil, :a, (:arglist))) ]

goto    319
        [ 0 19 226 54 319 ]

/home/alex/Programming/Ruby/opal/lib/opal/parser.rb:81:in `on_error':  (RuntimeError)
parse error on value "}" (tRCURLY) :(string):1
Source: (string):1:11
return %{a}
~~~~~~~~~~^
        from /home/alex/.rbenv/versions/2.2.2/lib/ruby/2.2.0/racc/parser.rb:258:in `_racc_do_parse_c'
        from /home/alex/.rbenv/versions/2.2.2/lib/ruby/2.2.0/racc/parser.rb:258:in `do_parse'
        from /home/alex/Programming/Ruby/opal/lib/opal/parser.rb:53:in `parse_to_sexp'
        from /home/alex/Programming/Ruby/opal/lib/opal/parser.rb:38:in `parse'
        from test.rb:5:in `<main>'
```

**Comment by alexdowad** [comment]
Compared with the "good" example:

```
22:11 ~prog/Ruby/opal % nano test.rb                                           [master]
22:12 ~prog/Ruby/opal % cat test.rb                                            [master]
$:.unshift Dir.pwd+'/lib'
require 'opal'
p = Opal::Parser.new
p.instance_eval { @yydebug = true }
p p.parse('%{a}')
22:12 ~prog/Ruby/opal % ruby test.rb                                           [master]
read    :tSTRING_BEG(tSTRING_BEG) ["%{", [1, 0]]

shift   tSTRING_BEG
        [ (tSTRING_BEG ["%{", [1, 0]]) ]

goto    73
        [ 0 73 ]

reduce  <none> --> none
        [ (tSTRING_BEG ["%{", [1, 0]]) (none nil) ]

goto    385
        [ 0 73 385 ]
```

The problem is already apparent. Do you see the different with the "bad" example?

**Comment by alexdowad** [comment]
Here is the relevant lexer code:

```ruby
        elsif scan(/\%/)
          if scan(/\=/)
            @lex_state = :expr_beg
            return new_op_asgn('%')
          elsif check(/[^\s]/)
            if @lex_state == :expr_beg or (@lex_state == :expr_arg && @space_seen)
              start_word  = scan(/./)
              end_word    = { '(' => ')', '[' => ']', '{' => '}' }[start_word] || start_word
              self.strterm = new_strterm2(STR_DQUOTE, end_word, start_word)
              return :tSTRING_BEG
            end
          end

          self.set_arg_state

          return :tPERCENT
```

This is the same code which is emitting `tSTRING_BEG` in one case, and `tPERCENT` in the other. Do you see the conditional which makes the difference?

**Comment by alexdowad** [comment]
`@lex_state` is `:expr_mid` after lexing `return`.

Comparison with `parser` (it's starting to feel like cheating) suggests that `%{` should be lexed as `tSTRING_BEG` in `expr_mid` state.

**Comment by alexdowad** [comment]
Just opening a PR which fixes this.

**Comment by Papierkorb** [comment]
@alexdowad  :+1: Thanks!

**Comment by elia** [comment]
fixed by #1225
